### PR TITLE
Moved efi to constructor

### PR DESCRIPTION
**Pull request by mnilsson23**

### CODE REVIEW

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java [hunk 1]
```diff
@@ -75,11 +75,14 @@
   private Semaphore querySemaphore; // limits the number of threads per query, so that multiple requests can be serviced simultaneously
 
   public ModelQuery(LTRScoringModel meta) {
-    this(meta, false);
+    this(meta, null, false);
```
Comment (diegoceccarelli) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java]: what about using `Collections.emptyMap()`, instead of `null` ?

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java [hunk 2]
```diff
@@ -104,10 +107,6 @@ public void setOriginalQuery(Query mainQuery) {
     originalQuery = mainQuery;
   }
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java]: `efi` can be marked `final` now then (no need to update this PR for just that though, will include in a future PR)

File: solr/contrib/ltr/src/test/org/apache/solr/ltr/ranking/TestModelQuery.java [hunk 3]
```diff
@@ -141,28 +141,28 @@ public void testModelQueryEquality() throws ModelException {
         "testModelName",
         features, norms, "testStoreName", allFeatures, modelParams);
 
-    final ModelQuery m1 = new ModelQuery(algorithm1);
-    final ModelQuery m2 = new ModelQuery(algorithm1);
-
-    assertEquals(m1, m2);
-    assertEquals(m1.hashCode(), m2.hashCode());
+    final ModelQuery m0 = new ModelQuery(algorithm1);
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/test/org/apache/solr/ltr/ranking/TestModelQuery.java]: observation: unrelated change?

File: solr/contrib/ltr/src/test/org/apache/solr/ltr/ranking/TestModelQuery.java [hunk 4]
```diff
@@ -141,28 +141,28 @@ public void testModelQueryEquality() throws ModelException {
...
+    final ModelQuery m2 = new ModelQuery(algorithm1, externalFeatureInfo2, false);
 
+
+    // Models with same algorithm and efis, just in different order should be the same
     assertEquals(m1, m2);
     assertEquals(m1.hashCode(), m2.hashCode());
 
+    // Models with same algorithm, but different efi content should not match
+    assertFalse(m1.equals(m0));
+    assertFalse(m1.hashCode() == m0.hashCode());
```
Comment (diegoceccarelli) [inline solr/contrib/ltr/src/test/org/apache/solr/ltr/ranking/TestModelQuery.java]: I think `assertFalse(m1.hashCode() == m0.hashCode());` it is useless because by contract If two objects are equal then they must have the same hash code.

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java (hunk 1 above)
Comment (mnilsson23) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java]: The previous code already had efi (un)initialized to null.  The existing code checks for null efis, so this is consistent with the previous behavior.

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java (hunk 2 above)
Comment (mnilsson23) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java]: yeah, we can do a final pass through for formatting and finals at the end

File: solr/contrib/ltr/src/test/org/apache/solr/ltr/ranking/TestModelQuery.java (hunk 3 above)
Comment (mnilsson23) [inline solr/contrib/ltr/src/test/org/apache/solr/ltr/ranking/TestModelQuery.java]: No, the old code checked the same 2 objects and their differences when calling setEfi.  Since now it is part of the constructor, I had to make a 3rd object without efi to do comparisons against.

File: solr/contrib/ltr/src/test/org/apache/solr/ltr/ranking/TestModelQuery.java (hunk 4 above)
Comment (mnilsson23) [inline solr/contrib/ltr/src/test/org/apache/solr/ltr/ranking/TestModelQuery.java]: Correct, by contract they should have the same hash codes.  Since we are the ones writing our own equals and hash code functions, this unit test is asserting that the contract is enforced.

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java (hunk 1 above)
Comment (diegoceccarelli) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java]: so if we add empty map we can remove the check on null?
